### ISSUE TITLE
Confusing tooltip text for field

### ISSUE THREAD

**Issue by DanTup** [Dart-Code/Dart-Code#555]
Using this revision:

https://github.com/DanTup/cambridge/commit/229c121ada718cbe88eed3f1bfea44ee6dd263f1

If you hover over `displayFrame` in `_SpectrumState.createSpectrumFrame` the tooltip says `_SpectrumState Image displayFrame`. Seems a bit weird having the enclosing type at the start on the same line as the actual type (why doesn't it just say `Image displayFrame` or even `Image`?). Not sure if it's showing too much info, or just badly formatted.

**Comment by DanTup** [comment]
@devoncarew Looks like this is from your original implementation:

https://github.com/Dart-Code/Dart-Code/commit/a612adc2d71178f122509407a776ad075253e109#diff-4383d7ce14d3ff4de58630d4092f1bd9R46

It's currently rendering like this (for the example given above):

![Tooltip](https://user-images.githubusercontent.com/1078012/36090791-92ab085a-0fd9-11e8-9bac-dbb0ca17a940.png)

I think the confusing thing (to me) is the space instead of a dot. Do you know if this was deliberate (if so, what's the reasoning)? It looks deliberate because it uses a dot for `callable` but space for `field`:

```ts
if (containingClassDescription && callable) displayString += containingClassDescription + ".";
if (containingClassDescription && field) displayString += containingClassDescription + " ";
```

Here's what a TypeScript equiv looks like (which does use a dot):

![TS Tooltip](https://user-images.githubusercontent.com/1078012/36090851-bdd2b5c8-0fd9-11e8-8a77-d92ce020f81a.png)

I think a dot makes more sense in all cases but I don't want to just change it without understanding.

**Comment by DanTup** [comment]
Actually, just changing to a dot won't fix it, since then it'd display `_SpectrumState.Image displayFrame` but probably it should be `Image _SpectrumState.displayFrame` (assuming we want to show both the type of the field and the containing type).

**Comment by devoncarew** [comment]
Not sure this helps, but for dartdoc, we chose to display the types after the name of the symbol (reverse from how they're declared). https://api.dartlang.org/stable/1.24.3/dart-math/dart-math-library.html

I don't know that this is an improvement here, but the equivalent would be `_SpectrumState.displayFrame -> Image` or `_SpectrumState.displayFrame: Image`.

Sounds like that's similar to what TS does.

**Comment by DanTup** [comment]
`_SpectrumState.displayFrame -> Image` seems reasonable to me (Dart Code users see that format in some other places anyway). The colon format I think woul dbe weird in Dart - works in TS because that's what the declarations look like.

I might take this opportunity to see if I can get some automated tests calling this provider; since there are a lot of different types of data returned here it'd be easy for me to make a change that would fix this case and negatively affect some others.

**Comment by DanTup** [comment]
Automated tests are kinda blocked by https://github.com/Microsoft/vscode/issues/43581 (seems like I must be doing something wrong, but can't figure it out) so will just have to change and test carefully manually.

**Comment by DanTup** [comment]
This needs more thought, as we don't actually have all the info we need to display this as I'd hoped. For example the hover info for a getter looks like this:

```js
{
	"id": "78",
	"result": {
		"hovers": [
			{
				"offset": 111,
				"length": 8,
				"containingLibraryPath": "c:\\Users\\danny\\Desktop\\DartHelloWorld\\bin\\main.dart",
				"containingLibraryName": "",
				"containingClassDescription": "Danny",
				"elementDescription": "get myGetter → String",
				"elementKind": "getter",
				"isDeprecated": false,
				"propagatedType": "String"
			}
		]
	}
}
```

The name (`myGetter`) only appears inside other strings, so there's (nice) way to show it that isn't part of the string `get myGetter → String` so putting the class name before the name won't work.

Maybe we should put the class name on its own line, so instead of being:

`MyClass get myGetter → String`

It's

```text
MyClass
get myGetter → String
```

Will definitely feel more comfortable doing this with tests covering everything; especially as there's a bunch of stuff I don't understand - for ex. why we have duplicated types so often (propagated type):

![Hover](https://user-images.githubusercontent.com/1078012/36207516-1e8885f4-118e-11e8-8ac0-3ead2d05a432.png)

**Comment by DanTup** [comment]
Automated tests are working now; we have to pass the folder as an argument to launching the test run (since opening a folder once inside a test causes a reload and the test code gets aborted).

Should be able to put some tests around this functionality now so we can see the impact of changing things.

**Comment by DanTup** [comment]
Turns out this was even worse than I noticed - we get the class names doubled for classes you're not inside: `Future.Future.then`!

So, I'm just dropping the class name here. Seems like the server is already including the class name in `elementDescription` when it's useful and all the other cases seem to be obvious what the class is (eg. you're already inside it, of it's fully qualified).
